Python/compile.c

```diff
@@ -5375,26 +5383,148 @@ compiler_async_comprehension_generator(struct compiler *c, location loc,
+                if (state->temp_symbols == NULL) {
+                    state->temp_symbols = PyDict_New();
+                    if (state->temp_symbols == NULL) {
+                        return ERROR;
+                    }
+                }
+                // update the symbol to the in-comprehension version and save
+                // the outer version; we'll restore it after running the
+                // comprehension
+                Py_INCREF(outv);
+                if (PyDict_SetItem(c->u->u_ste->ste_symbols, k, v)) {
+                    return ERROR;
+                }
+                if (PyDict_SetItem(state->temp_symbols, k, outv)) {
+                    return ERROR;
+                }
+                Py_DECREF(outv);
+            } else if (scope == LOCAL || scope == CELL) {
+                // local names bound in comprehension must be isolated from
+                // outer scope; push existing value (which may be NULL if
+                // not defined) on stack
+                if (state->pushed_locals == NULL) {
+                    state->pushed_locals = PyList_New(0);
+                    if (state->pushed_locals == NULL) {
+                        return ERROR;
+                    }
+                }
+                // in the case of a cell, this will actually push the cell
+                // itself to the stack, then we'll create a new one for the
+                // comprehension and restore the original one after
+                ADDOP_NAME(c, loc, LOAD_FAST_OR_NULL, k, varnames);
+                if (scope == CELL) {
+                    ADDOP_NAME(c, loc, MAKE_CELL, k, cellvars);
+                }
+                if (PyList_Append(state->pushed_locals, k)) {
+                    return ERROR;
+                }
+            }
+        }
+    }
+    return SUCCESS;
+}
+
+static int
+pop_inlined_comprehension_state(struct compiler *c, location loc,
+                                inlined_comprehension_state state)
+{
+    PyObject *k, *v;
+    Py_ssize_t pos = 0;
+    if (state.temp_symbols) {
+        // restore scope for globals that we temporarily set as locals
+        while (PyDict_Next(state.temp_symbols, &pos, &k, &v)) {
+            if (PyDict_SetItem(c->u->u_ste->ste_symbols, k, v)) {
+                return ERROR;
+            }
+        }
+    }
+    if (state.pushed_locals) {
+        // pop names we pushed to stack earlier
+        for (Py_ssize_t i = PyList_GET_SIZE(state.pushed_locals); i > 0; --i) {
+            k = PyList_GetItem(state.pushed_locals, i - 1);
+            if (k == NULL) {
+                return ERROR;
+            }
+            // preserve the list/dict/set result of the comprehension as TOS; we
+            // could be cleverer about minimizing swaps but it doesn't matter
+            // because `apply_static_swaps` will eliminate all of these anyway
+            ADDOP_I(c, loc, SWAP, 2);
+            ADDOP_NAME(c, loc, STORE_FAST, k, varnames);
```

Comment:
We could be storing `NULL` here. That is OK for the interpreter, but the dataflow analysis in `add_checks_for_loads_of_uninitialized_variables` in the compiler assumes that `STORE_FAST` leaves the variable initialized.

What you'll need to do is add a pseudo instruction, and convert that to `STORE_FAST` in the assembler. That way the dataflow analysis won't be fooled.

You might want to update `add_checks_for_loads_of_uninitialized_variables` to understand the `LOAD_FAST_AND_CLEAR`, `STORE_FAST_MAYBE_NULL` pairs, but that should wait for another PR.

Reply:
Here is a test case:
```Python
def f():
     if False:
          x = 0
     [x for _ in [1]]
     return x
```

Reply:
Good call, thanks!

The test case needs to use `[x for x in [1]]`, not `[x for _ in [1]]` -- in the latter case, `x` is not bound in the comprehension and is just a free var (converted to local by inlining) so we don't push/pop it at all.